Include AuthProperties in OIDC RedirectContext

HaoK opened the pull request:
cc @Tratcher

dnfclas commented:
Hi __@HaoK__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

Tratcher commented on src/Microsoft.AspNet.Authentication.OpenIdConnect/Events/RedirectContext.cs:
```diff
@@ -12,9 +13,12 @@ namespace Microsoft.AspNet.Authentication.OpenIdConnect
     public class RedirectContext : BaseOpenIdConnectContext
     {
-        public RedirectContext(HttpContext context, OpenIdConnectOptions options)
+        public RedirectContext(HttpContext context, OpenIdConnectOptions options, AuthenticationProperties properties)
```
Don't put it in the constructor, the property is already settable.

HaoK replied:
Updated, to use property

Tratcher commented:
This ask is generally applicable to all middleware events that can be triggered from Challenge, SignIn, SignOut, etc.. No?

HaoK commented:
Looks like that's only one other OIDC event, the AuthorizationCodeReceivedContext

HaoK commented:
I guess we can flow it everywhere that we have properties, can add it to the ValidatedContext too

HaoK commented:
Worth considering moving this to the base context, as there's only three context's that don't set this now

Tratcher commented:
How about the non-OIDC middleware?

HaoK commented:
OAuth looks to already have these in the base events, JWT most likely has the same issues I can update that too

HaoK commented:
Changes like this really make me think the events stuff is way too complicated :)

PinpointTownes commented:
Definitely a must :clap:

Tratcher commented:
CookieRedirectContext?

HaoK commented:
Updated Cookies, added Properties to SigningOutContext and the CookieRedirectContext.  We use the auth ticket's properties for the FinishResponse method since we don't have any other source for the properties there...

PinpointTownes commented on src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationHandler.cs:
```diff
@@ -279,6 +279,7 @@ private CookieOptions BuildCookieOptions()
             var context = new CookieSigningOutContext(
                 Context,
                 Options,
+                new AuthenticationProperties(signOutContext.Properties),
```
Any reason to flow the properties using a constructor parameter here and a property everywhere else?

Personally, I'd use a parameter everywhere (and probably make the `Properties` property get-only).

HaoK replied:
Dev style mostly, I prefer ctor.  I think @Tratcher preferred properties.  I can make them consistent and ctor required everywhere I guess.  Chris thoughts?

Tratcher replied:
Property

HaoK replied:
Property made sense when they are optional, if we are making them required, i think it belongs in the ctor.

HaoK replied:
compile errors >> null ref exceptions...

PinpointTownes replied:
Internal then? Because having a settable property worries me a bit.

(and it wouldn't be very consistent: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.Cookies/Events/CookieSignedInContext.cs#L49)

Tratcher replied:
No, it's not internal. We don't care if anyone changes it, we won't flow it.

HaoK replied:
Why is this one second class compared to the other ctor arguments which could be left as get/set properties as well

PinpointTownes replied:
No, it's not internal, it's private and set from the constructor :trollface: 
Hence my remark.

> We don't care if anyone changes it, we won't flow it.

Not sure why you want to let the user think he's allowed to replace the instance, then.

HaoK replied:
CookieSigningInContext for example takes several properties in the ctor and has get/set for each, I would think its either or, public set (no ctor), or private set in ctor, with strong preference towards private/ctor IMO

Tratcher replied:
Required params go in the constructor (especially if they're passed to a base constructor). optional params use properties.

HaoK replied:
Exactly, the point of this change was to make AuthProperties required no?

PinpointTownes replied:
Curious: why do you consider them as optional?

Tratcher replied:
It's extra data, one less thing to set up for tests.

HaoK replied:
I'm not sure that's a valid argument right now, since we don't test any of these contexts anyways :)  I'd say making them not settable at all, is actually less surface area in the long term anyways...

PinpointTownes commented on src/Microsoft.AspNet.Authentication.Cookies/Events/CookieRedirectContext.cs:
```diff
@@ -29,5 +30,7 @@ public CookieRedirectContext(HttpContext context, CookieAuthenticationOptions op
         [SuppressMessage("Microsoft.Design", "CA1056:UriPropertiesShouldNotBeStrings", Justification = "Represents header value")]
         public string RedirectUri { get; set; }
+
+        public AuthenticationProperties Properties { get; set; }
```
nit: no documentation love for these properties in this PR? :trollface:

HaoK replied:
My plan was to wait and do a full pass, lots of doc comments are missing/out of date currently in this repo...  This issue is tracking all up https://github.com/aspnet/Security/issues/190

muratg commented:
@HaoK @Tratcher is this ready?

HaoK commented:
Just a minor remaining stylistic/consistency thing I need to hash out with @Tratcher but this is basically done yes...

muratg commented:
Cool. I think @Tratcher will be out till January. Let's get it in soon after that. :)

HaoK commented:
Discussed with @Tratcher he is ok if we go with required ctor as long as there is only one ctor per context

muratg commented:
:+1:

HaoK commented:
Updated, promoted AuthenticationProperties to a required ctor argument in contexts where its used.  I left the Cookie properties settable as those are intended to be used to replace them in some place.  But I tried to make them get only everywhere else

Tratcher commented on src/Microsoft.AspNet.Authentication.Google/GoogleHandler.cs:
```diff
@@ -31,12 +31,8 @@ public GoogleHandler(HttpClient httpClient)
-                Principal = new ClaimsPrincipal(identity)
-            };
+
+            var context = new OAuthCreatingTicketContext(new ClaimsPrincipal(identity), properties, Context, Options, Backchannel, tokens, payload);
```
This constructor is getting a bit out of hand.

HaoK replied:
Yeah its seems big, but I would assume all of these are required for OAuth ticket contexts?  Making them properties doesn't really change anything except moving them into a { } block and making it possible to forget to set one...

PinpointTownes replied:
We could replace the `ClaimsPrincipal` and `AuthenticationProperties` arguments by a single `AuthenticationTicket`.

PinpointTownes replied:
Also, since the last parameter is optional, you could replace it a by property.

HaoK replied:
Doesn't really seem worth the additional churn at this point as it doesn't seem to buy us that much does it?

HaoK replied:
The user overload ctor?  Sure I can fix that

Tratcher replied:
Actually, passing in a ticket would make sense, that's the ultimate output at the end of this method.

HaoK replied:
I'll do that in a followup PR then, as its basically moving the creation of the ticket from the method CreateTicketAsync into whatever is provided in the context

Tratcher commented:
:shipit:

HaoK commented:
3257a82367e946a351cc1eaa02a9bb5dae2a432b
